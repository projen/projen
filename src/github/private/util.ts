export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
