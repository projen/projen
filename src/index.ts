export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
