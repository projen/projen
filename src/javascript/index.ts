export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
