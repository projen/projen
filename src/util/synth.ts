import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import * as JSONC from "comment-json";
import { CommentArray } from "comment-json";
import * as glob from "fast-glob";
import { JsonFile } from "../json";
import { Project } from "../project";

/**
 * Options for the Snapshot synthesis
 */
export interface SnapshotOptions {
  /**
   * Parse .json files as a JS object for improved inspection.
   * This will fail if the contents are invalid JSON.
   *
   * @default true parse .json files into an object
   */
  readonly parseJson?: boolean;
}

export interface SynthOutput {
  [filePath: string]: any;
}

/**
 * Creates a snapshot of the files generated by a project. Ignores any non-text
 * files so that the snapshots are human readable.
 */
export function synthSnapshot(
  project: Project,
  options: SnapshotOptions = {}
): SynthOutput {
  // defensive: verify that "outdir" is actually in a temporary directory
  if (
    !path.resolve(project.outdir).startsWith(os.tmpdir()) &&
    !project.outdir.includes("project-temp-dir")
  ) {
    throw new Error(
      "Trying to capture a snapshot of a project outside of tmpdir, which implies this test might corrupt an existing project"
    );
  }

  const synthed = Symbol.for("synthed");
  if (synthed in project) {
    throw new Error("duplicate synth()");
  }

  (project as any)[synthed] = true;

  const ENV_PROJEN_DISABLE_POST = process.env.PROJEN_DISABLE_POST;
  try {
    process.env.PROJEN_DISABLE_POST = "true";
    project.synth();
    const ignoreExts = ["png", "ico"];
    return directorySnapshot(project.outdir, {
      ...options,
      excludeGlobs: ignoreExts.map((ext) => `**/*.${ext}`),
      supportJsonComments: project.files.some(
        // At least one json file in project supports comments
        (file) => file instanceof JsonFile && file.supportsComments
      ),
    });
  } finally {
    fs.rmSync(project.outdir, { force: true, recursive: true });

    // values assigned to process.env.XYZ are automatically converted to strings
    if (ENV_PROJEN_DISABLE_POST === undefined) {
      delete process.env.PROJEN_DISABLE_POST;
    } else {
      process.env.PROJEN_DISABLE_POST = ENV_PROJEN_DISABLE_POST;
    }
  }
}

export interface DirectorySnapshotOptions extends SnapshotOptions {
  /**
   * Globs of files to exclude.
   * @default [] include all files
   */
  readonly excludeGlobs?: string[];

  /**
   * Only snapshot the names of files and not their contents.
   * The value for a path will be `true` if it exists.
   *
   * @default false include file content
   */
  readonly onlyFileNames?: boolean;

  /**
   * Parses files with different parser, supporting comments
   * inside .json files.
   * @default false
   */
  readonly supportJsonComments?: boolean;
}

function isJsonLikeFile(filePath: string): boolean {
  const file = filePath.toLowerCase();
  return (
    file.endsWith(".json") || file.endsWith(".json5") || file.endsWith(".jsonc")
  );
}

export function directorySnapshot(
  root: string,
  options: DirectorySnapshotOptions = {}
) {
  const output: SynthOutput = {};

  const files = glob.sync("**", {
    ignore: [".git/**", ...(options.excludeGlobs ?? [])],
    cwd: root,
    onlyFiles: true,
    followSymbolicLinks: false,
    dot: true,
  }); // returns relative file paths with POSIX separators

  const parseJson = options.parseJson ?? true;

  for (const file of files) {
    const filePath = path.join(root, file);

    let content;
    if (!options.onlyFileNames) {
      content = fs.readFileSync(filePath, "utf-8");
      if (parseJson && isJsonLikeFile(filePath)) {
        content = cleanCommentArrays(
          JSONC.parse(content, undefined, !options.supportJsonComments)
        );
      }
    } else {
      content = true;
    }

    output[file] = content;
  }

  return output;
}

/**
 * Converts type "CommentArray" back to regular JS "Array"
 * if there are no comments stored in it.
 * Prevents strict checks from failing.
 */
function cleanCommentArrays(obj: any): typeof obj {
  if (Array.isArray(obj) || isCommentArrayWithoutComments(obj)) {
    return Array.from(obj).map(cleanCommentArrays);
  }

  if (obj instanceof Object) {
    for (const p of Object.keys(obj)) {
      if (isCommentArrayWithoutComments(obj[p])) {
        obj[p] = Array.from(obj[p]).map(cleanCommentArrays);
      } else if (obj[p] instanceof Object) {
        obj[p] = cleanCommentArrays(obj[p]);
      }
    }
  }

  return obj;
}

/**
 * Checks if a "CommentArray" has no comments stored in it.
 */
function isCommentArrayWithoutComments(obj: any): boolean {
  return (
    obj instanceof CommentArray &&
    Object.getOwnPropertySymbols(obj).length === 0
  );
}
