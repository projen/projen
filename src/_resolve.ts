import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
