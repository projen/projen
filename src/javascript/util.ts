import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
