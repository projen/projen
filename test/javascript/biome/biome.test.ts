import { Biome, biome_config } from "../../../src/javascript";
import { NodeProject } from "../../../src/javascript/node-project";
import {
  TypeScriptProject,
  type TypeScriptProjectOptions,
} from "../../../src/typescript";
import { synthSnapshot } from "../../../src/util/synth";

const getTestProject = (
  projenOptions: Partial<TypeScriptProjectOptions>
): TypeScriptProject => {
  return new TypeScriptProject({
    name: "test-project",
    defaultReleaseBranch: "main",
    sampleCode: false,
    githubOptions: {
      mergify: false,
    },
    ...projenOptions,
  });
};

describe("biome", () => {
  describe("with default configuration", () => {
    describe("generates correct snapshot for", () => {
      const project = getTestProject({ eslint: false, prettier: false });
      new Biome(project);

      const snapshots = synthSnapshot(project);
      for (const file of Object.keys(snapshots)) {
        it(file, () => {
          expect(snapshots[file]).toMatchSnapshot();
        });
      }
    });
  });

  describe("with", () => {
    it("formatter enabled", () => {
      const project = getTestProject({ eslint: false, prettier: false });
      new Biome(project, {
        formatter: true,
      });

      const config: biome_config.IConfiguration =
        synthSnapshot(project)["biome.jsonc"];
      expect(config.formatter?.enabled).toBeTruthy();
    });

    it("organizer enabled", () => {
      const project = getTestProject({ eslint: false, prettier: false });
      new Biome(project, {
        organizeImports: true,
      });

      const config: biome_config.IConfiguration =
        synthSnapshot(project)["biome.jsonc"];
      expect(config.organizeImports?.enabled).toBeTruthy();
    });

    it("custom overrides", () => {
      const project = getTestProject({ eslint: false, prettier: false });
      new Biome(project, {
        formatter: true,
        biomeConfig: {
          files: {
            ignore: ["ignored-file.txt"],
          },
        },
      });

      const config: biome_config.IConfiguration =
        synthSnapshot(project)["biome.jsonc"];
      expect(config.formatter?.enabled).toBeTruthy();
      expect(config.files?.ignore).toContain("ignored-file.txt");
    });

    test("additional lint pattern", () => {
      // GIVEN
      const project = new NodeProject({
        name: "test",
        defaultReleaseBranch: "master",
        prettier: false,
      });

      // WHEN
      const biome = new Biome(project, {});

      const taskStep = biome.task.steps[0];
      const newTestArg = "--foo";
      biome.task.reset(taskStep.exec, { args: [newTestArg] });

      console.dir(biome.task.steps[0]);
      const newLintPattern = "bar";
      biome.addLintPattern(newLintPattern);

      // THEN
      expect(biome.task.steps[0].args).toContain(newTestArg);
      expect(biome.task.steps[0].exec).toContain(newLintPattern);
    });
  });

  describe("of", () => {
    it("returns Biome instance when found", () => {
      const project = getTestProject({ eslint: false, prettier: false });
      const biome = new Biome(project);

      const found = Biome.of(project);
      expect(found).toBe(biome);
    });

    it("returns undefined when not found", () => {
      const project = getTestProject({ eslint: false, prettier: false });
      const found = Biome.of(project);
      expect(found).toBeUndefined();
    });
  });
});

describe("Configuration", () => {
  describe("should match snapshot", () => {
    describe("without overrides", () => {
      it("without options", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, {});
        const config = synthSnapshot(project)["biome.jsonc"];
        expect(config).toMatchSnapshot();
      });

      it("with linter", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, { linter: true });
        const config = synthSnapshot(project)["biome.jsonc"];
        expect(config).toMatchSnapshot();
      });

      it("with formatter", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, { formatter: true });
        const config = synthSnapshot(project)["biome.jsonc"];
        expect(config).toMatchSnapshot();
      });

      it("with import organizer", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, { organizeImports: true });
        const config = synthSnapshot(project)["biome.jsonc"];
        expect(config).toMatchSnapshot();
      });

      it("with all features", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, {
          linter: true,
          formatter: true,
          organizeImports: true,
        });
        const config = synthSnapshot(project)["biome.jsonc"];
        expect(config).toMatchSnapshot();
      });
    });
  });

  describe("should", () => {
    describe("merge arrays", () => {
      it("when mergeArraysInConfiguration is true", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, {
          linter: true,
          mergeArraysInConfiguration: true,
          biomeConfig: {
            linter: {
              ignore: ["**/bin"],
            },
          },
        });
        const mergedConfig = synthSnapshot(project)["biome.jsonc"];

        expect(mergedConfig.linter?.ignore?.length >= 2).toBe(true);
      });

      it("without duplicate values", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, { linter: true });
        const config = synthSnapshot(project)["biome.jsonc"];

        expect(config.linter?.ignore?.[0]).toBeDefined();

        const project2 = getTestProject({ eslint: false, prettier: false });
        new Biome(project2, {
          linter: true,
          biomeConfig: {
            linter: {
              ignore: [...config.linter!.ignore!],
            },
          },
        });
        const mergedConfig = synthSnapshot(project2)["biome.jsonc"];

        expect(mergedConfig.linter!.ignore!.length).toBe(
          config.linter!.ignore!.length
        );
      });
    });

    describe("replace whole array", () => {
      it("when mergeArraysInConfiguration is false", () => {
        const project = getTestProject({ eslint: false, prettier: false });
        new Biome(project, {
          linter: true,
          mergeArraysInConfiguration: false,
          biomeConfig: {
            linter: {
              ignore: ["foo"],
            },
          },
        });
        const mergedConfig = synthSnapshot(project)["biome.jsonc"];

        expect(mergedConfig.linter?.ignore).toEqual(["foo"]);
      });
    });

    it("make deep merge", () => {
      const project = getTestProject({ eslint: false, prettier: false });
      new Biome(project, {
        linter: true,
        biomeConfig: {
          linter: {
            rules: {
              a11y: {
                noBlankTarget: "error",
              },
            },
          },
        },
      });
      const mergedConfig = synthSnapshot(project)["biome.jsonc"];

      expect(mergedConfig.linter!.rules).toBeDefined();
      expect(mergedConfig.linter!.rules).toMatchObject({
        recommended: true,
        a11y: {
          noBlankTarget: "error",
        },
      });
    });
  });
});
