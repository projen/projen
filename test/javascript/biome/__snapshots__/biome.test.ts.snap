// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration should match snapshot without overrides with all features 1`] = `
{
  "$schema": "https://biomejs.dev/schemas/latest/schema.json",
  "assist": {
    "actions": {
      "recommended": true,
    },
    "enabled": true,
  },
  "files": {
    "ignoreUnknown": false,
    "includes": [
      ".projenrc.js",
    ],
  },
  "formatter": {
    "enabled": true,
    "indentStyle": "tab",
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "double",
    },
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
    },
  },
  "vcs": {
    "clientKind": "git",
    "enabled": true,
    "useIgnoreFile": true,
  },
}
`;

exports[`Configuration should match snapshot without overrides with assist 1`] = `
{
  "$schema": "https://biomejs.dev/schemas/latest/schema.json",
  "assist": {
    "actions": {
      "recommended": true,
    },
    "enabled": true,
  },
  "files": {
    "ignoreUnknown": false,
    "includes": [
      ".projenrc.js",
    ],
  },
  "formatter": {
    "enabled": true,
    "indentStyle": "tab",
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "double",
    },
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
    },
  },
  "vcs": {
    "clientKind": "git",
    "enabled": true,
    "useIgnoreFile": true,
  },
}
`;

exports[`Configuration should match snapshot without overrides with formatter 1`] = `
{
  "$schema": "https://biomejs.dev/schemas/latest/schema.json",
  "assist": {
    "actions": {
      "recommended": true,
    },
    "enabled": true,
  },
  "files": {
    "ignoreUnknown": false,
    "includes": [
      ".projenrc.js",
    ],
  },
  "formatter": {
    "enabled": true,
    "indentStyle": "tab",
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "double",
    },
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
    },
  },
  "vcs": {
    "clientKind": "git",
    "enabled": true,
    "useIgnoreFile": true,
  },
}
`;

exports[`Configuration should match snapshot without overrides with linter 1`] = `
{
  "$schema": "https://biomejs.dev/schemas/latest/schema.json",
  "assist": {
    "actions": {
      "recommended": true,
    },
    "enabled": true,
  },
  "files": {
    "ignoreUnknown": false,
    "includes": [
      ".projenrc.js",
    ],
  },
  "formatter": {
    "enabled": true,
    "indentStyle": "tab",
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "double",
    },
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
    },
  },
  "vcs": {
    "clientKind": "git",
    "enabled": true,
    "useIgnoreFile": true,
  },
}
`;

exports[`Configuration should match snapshot without overrides without options 1`] = `
{
  "$schema": "https://biomejs.dev/schemas/latest/schema.json",
  "assist": {
    "actions": {
      "recommended": true,
    },
    "enabled": true,
  },
  "files": {
    "ignoreUnknown": false,
    "includes": [
      ".projenrc.js",
    ],
  },
  "formatter": {
    "enabled": true,
    "indentStyle": "tab",
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "double",
    },
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
    },
  },
  "vcs": {
    "clientKind": "git",
    "enabled": true,
    "useIgnoreFile": true,
  },
}
`;

exports[`biome with default configuration generates correct snapshot for biome.jsonc 1`] = `
{
  "$schema": "https://biomejs.dev/schemas/latest/schema.json",
  "assist": {
    "actions": {
      "recommended": true,
    },
    "enabled": true,
  },
  "files": {
    "ignoreUnknown": false,
    "includes": [
      "src/**",
      "test/**",
      ".projenrc.ts",
      "projenrc/**",
    ],
  },
  "formatter": {
    "enabled": true,
    "indentStyle": "tab",
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "double",
    },
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
    },
  },
  "vcs": {
    "clientKind": "git",
    "enabled": true,
    "useIgnoreFile": true,
  },
}
`;
